File: c++/puzzle3-2.cpp
#include <iostream>
#include <string>
#include <string_view>
#include <vector>
#include <utility>
#include <cassert>
#include <charconv>
#include <cctype>
#include <functional>

struct Number {
    std::size_t start;
    std::size_t end;
    unsigned long val;
};

struct Symbol {
    std::size_t start;
    char ch;
};

using SchematicLine = std::pair<std::vector<Number>, std::vector<Symbol>>;

SchematicLine parse_line(std::string_view line)
{
    std::vector<Number> nums;
    std::vector<Symbol> syms;
    auto start = line.data();
    while (!line.empty()) {
        auto pos = line.find_first_not_of('.');
        if (pos == line.npos)
            break;
        line.remove_prefix(pos);

        pos = line.data() - start;
        if (std::isdigit(line[0])) {
            unsigned long val = 0;
            auto [ptr, ec] = std::from_chars(line.data(),
                line.data() + line.size(), val);
            assert(ec == std::errc{} && val);
            std::size_t len = ptr - line.data();
            nums.push_back({pos, pos + len - 1, val});
            line.remove_prefix(len);
        } else {
            syms.push_back({pos, line[0]});
            line.remove_prefix(1);
        }
    }
    return {nums, syms};
}

unsigned long sum_part_numbers(const SchematicLine& prev,
                               const SchematicLine& cur,
                               const SchematicLine& next)
{
    using Ref = std::reference_wrapper<const SchematicLine>;
    Ref lines[] = {prev, cur, next};
    unsigned long sum = 0;
    for (auto& sym : cur.second) {
        if (sym.ch != '*')
            continue;
        unsigned long vals[2] = { 0 };
        int i = 0;
        for (auto& line : lines) {
            for (auto& num : line.get().first) {
                if (sym.start > 0 && num.end < sym.start - 1)
                    continue;
                if (num.start > sym.start + 1)
                    break;
                if (i < 2)
                    vals[i] = num.val;
                ++i;
                if (i > 2)
                    break;
            }
            if (i > 2)
                break;
        }
        if (i == 2)
            sum += vals[0] * vals[1];
    }
    return sum;
}

int main()
{
    SchematicLine prev, cur, next;
    unsigned long total = 0;
    for (std::string line; std::getline(std::cin, line); ) {
        next = parse_line(line);
        auto sum = sum_part_numbers(prev, cur, next);
        std::cout << sum << '\n';
        total += sum;
        prev = std::move(cur);
        cur = std::move(next);
    }
    auto sum = sum_part_numbers(prev, cur, SchematicLine{});
    std::cout << sum << '\n';
    total += sum;
    std::cout << "total: " << total << '\n';
    return 0;
}

File: rust/src/bin/puzzle3-2.rs
use atoi::FromRadix10;
use std::str::FromStr;

struct Number {
    value: u32,
    start_pos: usize,
    end_pos: usize,
}

struct Symbol {
    ch: u8,
    pos: usize,
}

struct SchematicLine {
    numbers: Vec<Number>,
    symbols: Vec<Symbol>,
}

impl SchematicLine {
    fn sum(&self, prev: Option<&Self>, next: Option<&Self>) -> u32 {
        let mut sum: u32 = 0;
        let lines: Vec<&SchematicLine> = [prev, Some(self), next]
            .into_iter()
            .filter_map(|x| x)
            .collect();
        'symbol: for sym in &self.symbols {
            if sym.ch != b'*' {
                continue;
            }
            let mut adjacent = Vec::with_capacity(2);
            for line in &lines {
                for num in &line.numbers {
                    if num.end_pos < sym.pos.saturating_sub(1) {
                        continue;
                    } else if num.start_pos > sym.pos + 1 {
                        break;
                    }
                    if adjacent.len() == 2 {
                        continue 'symbol;
                    }
                    adjacent.push(num.value);
                }
            }
            if adjacent.len() == 2 {
                sum += adjacent.iter().product::<u32>();
            }
        }
        sum
    }
}

impl FromStr for SchematicLine {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut numbers = Vec::new();
        let mut symbols = Vec::new();
        let bytes = s.as_bytes();
        let mut pos = 0;
        while let Some(&ch) = bytes.get(pos) {
            if ch == b'.' {
                pos += 1;
            } else if ch.is_ascii_digit() {
                let (value, len) = u32::from_radix_10(&bytes[pos..]);
                assert!(len > 0);
                numbers.push(Number {
                    value,
                    start_pos: pos,
                    end_pos: pos + len - 1,
                });
                pos += len;
            } else {
                assert!(ch.is_ascii_graphic());
                symbols.push(Symbol { ch, pos });
                pos += 1;
            }
        }
        Ok(SchematicLine { numbers, symbols })
    }
}

fn main() {
    let mut sum: u32 = 0;
    let mut lines = std::io::stdin().lines().map(|s| s.unwrap());
    let mut prev: Option<SchematicLine> = None;
    let mut cur: SchematicLine = lines.next().unwrap().parse().unwrap();
    let mut next: SchematicLine;
    for line in lines {
        next = line.parse().unwrap();
        sum += cur.sum(prev.as_ref(), Some(&next));
        prev = Some(cur);
        cur = next;
    }
    sum += cur.sum(prev.as_ref(), None);
    println!("{sum}");
}
